pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,640 @@
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&covev1client.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		if err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage("cronjob_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return nil, err
+		}
+	}
+
+	jm := &ControllerV2{
+		queue:    workqueue.NewNamedDelayingQueue("cronjob"),
+		recorder: eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "cronjob-controller"}),
+
+		jobControl:     realJobControl{KubeClient: kubeClient},
+		cronJobControl: &realCJControl{KubeClient: kubeClient},
+
+		jobLister:     jobInformer.Lister(),
+		cronJobLister: cronJobsInformer.Lister(),
+
+		jobListerSynced:     jobInformer.Informer().HasSynced,
+		cronJobListerSynced: cronJobsInformer.Informer().HasSynced,
+		now:                 time.Now,
+	}
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    jm.addJob,
+		UpdateFunc: jm.updateJob,
+		DeleteFunc: jm.deleteJob,
+	})
+
+	cronJobsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+		UpdateFunc: jm.updateCronJob,
+		DeleteFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+	})
+
+	return jm, nil
+}
+
+// Run starts the main goroutine responsible for watching and syncing jobs.
+func (jm *ControllerV2) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer jm.queue.ShutDown()
+
+	klog.Infof("Starting cronjob controller v2")
+	defer klog.Infof("Shutting down cronjob controller v2")
+
+	go jm.enqueueAllExistingCronJobs()
+
+	if !cache.WaitForNamedCacheSync("cronjob", stopCh, jm.jobListerSynced, jm.cronJobListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(jm.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+// enqueueAllExistingCronJobs enqueues all the cronjobs and is thread-safe.
+func (jm *ControllerV2) enqueueAllExistingCronJobs() {
+	cronjobs, err := jm.cronJobLister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("unable to list cronjobs at the start of cronjob controller, error: %v", err))
+		return
+	}
+	for _, cronjob := range cronjobs {
+		jm.enqueueController(cronjob)
+	}
+	return
+}
+
+func (jm *ControllerV2) worker() {
+	for jm.processNextWorkItem() {
+	}
+}
+
+func (jm *ControllerV2) processNextWorkItem() bool {
+	key, quit := jm.queue.Get()
+	if quit {
+		return false
+	}
+	defer jm.queue.Done(key)
+	requeueAfter, err := jm.sync(key.(string))
+	switch {
+	case err != nil:
+		utilruntime.HandleError(fmt.Errorf("error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		jm.queue.Add(key)
+	case requeueAfter != nil:
+		jm.queue.AddAfter(key, *requeueAfter)
+	default:
+		jm.queue.Done(key)
```

Comment:
What happens when you call Done() twice for the same thing - you have Done() already called in defer.

Reply:
It should be noop, but you're right it's not necessary.

Reply:
There could be a case where Done is called in differ even after the case executed by the switch added the key. In order to prevent that mishap seems like I should remove the differ call.